src/i2c.rs

```diff
@@ -256,202 +336,175 @@ impl AddressMode for SevenBitAddress {}
 
 impl AddressMode for TenBitAddress {}
 
+/// Direction of an I2C transfer.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum Direction {
+    /// I2C read (RnW bit is 1)
+    Read,
+    /// I2C write (RnW bit is 0)
+    Write,
+}
+
 /// Blocking I2C traits
 pub mod blocking {
+    use super::*;
 
-    use super::{AddressMode, ErrorType, SevenBitAddress};
-
-    /// Transactional I2C operation.
+    /// I2C device trait
     ///
-    /// Several operations can be combined as part of a transaction.
-    #[derive(Debug, PartialEq)]
-    pub enum Operation<'a> {
-        /// Read data into the provided buffer
-        Read(&'a mut [u8]),
-        /// Write data from the provided buffer
-        Write(&'a [u8]),
-    }
+    /// `I2cDevice` represents ownership over a single I2C device on a (possibly shared) bus.
+    ///
+    /// See the [module-level documentation](self) for important usage information.
+    pub trait I2cDevice: ErrorType {
+        /// I2C Bus type for this device.
+        type Bus: I2cBusBase;
```

Comment:
After seeing @eldruin's code, I think it is a bit unfortunate to see driver code requiring two bounds each time, like [here](https://github.com/eldruin/ads1x1x-rs/blob/aa05b219c22a13cb7291f3380e8ce13738eadfef/src/interface.rs#L20-L23):

```rust
impl<I2C, E> WriteData for I2cInterface<I2C>
where
    I2C: embedded_hal::i2c::blocking::I2cDevice<Error = E>,
    I2C::Bus: embedded_hal::i2c::blocking::I2cBus<Error = E>,
```

Unless I am completely off, this is caused by the way the `AddressMode` is implemented, right?  I wonder if there ever is a need to address a device in both seven-bit and ten-bit mode from the same `I2cDevice`, like the current implementation is structured to allow.

I think it might be better to make the I2cDevice also generic over the `AddressMode`:

```rust
pub trait I2cDevice<A: AddressMode = SevenBitAddress>: ErrorType {
    type Bus: I2cBus<A, Error = Self::Error>;
}
```

This should allow drivers to only require a single bound.

Reply:
I don't think this'd work, because all I2cDevice methods would be "duplicated" for each addr type. 

For example, if a driver needs to access both 7bit and 10bit addrs, it'd do

```rust
impl<I2C, E> WriteData for I2cInterface<I2C>
where
    I2C: I2cDevice<SevenBitAddress, Error = E> + I2cDevice<TenBitAddress, Error = E>,
```

but then when it does `bus.transaction(...)`, it'll be ambiguous whether it's `<bus as I2cDevice<SevenBitAddress>>::transaction()` or `<bus as I2cDevice<TenBitAddress>>::transaction()`, which makes no sense. Opening a transaction shouldn't care about the address mode.

Reply:
Also there'd be no guarantee `<BUS as I2cDevice<SevenBitAddress>>::Bus` and `<BUS as I2cDevice<TenBitAddress>>::BUS` are the same type, which is something we do want IMO.

Reply:
> For example, if a driver needs to access both 7bit and 10bit addrs, it'd do

My thought was that this usecase is exceedingly rare.  Maybe I am missing something but so far I have not seen any driver where this is required.